pennylane/gradients/pulse_generator_gradient.py

```diff
@@ -0,0 +1,664 @@
+    #    have been created for the pulse.
+    #    Correspondingly, these pointers define the sub-list of results that should be used
+    #    for the given pulse in the postprocessing function below.
+    #    The third entry in the value tuple contains the coefficients of the effective
+    #    generators of the given pulse. These are given in a tuple, with one entry per
+    #    parameter of the pulse.
+    #    For details on these coefficients, see _one_parameter_paulirot_coeffs.
+    cache = {"total_num_tapes": 0}
+    # ``gradient_data`` will contain tuples ``(int, int, tensor_like)``, with the first two
+    # entries being the start/end pointers explained above, and the third entry being the
+    # coefficients tensor for one particular parameter of a pulse (one entry of the coefficients
+    # tuple in the corresponding ``cache`` entry).
+    gradient_data = []
+    gradient_tapes = []
+    for idx, trainable_idx in enumerate(tape.trainable_params):
+        if trainable_idx not in argnum:
+            # Trainable parameters that are de-selected by ``argnum`` receive a vanishing
+            # gradient entry.
+            # Indicate that there are no tapes for this parameter by setting start==end
+            gradient_data.append((0, 0, None))
+            continue
+
+        # Generate new tapes and coefficients, or retrieve them from the cache, in case
+        # the pulse to which ``trainable_idx`` belongs was already treated.
+        # If new tapes and coefficients are created, they are added to the cache as well
+        tapes, data, cache = _generate_tapes_and_coeffs(tape, idx, atol, cache)
+
+        gradient_data.append(data)
+        gradient_tapes.extend(tapes)
+
+    # Extract some specifications about the original tape, required for output formatting
+    num_measurements = len(tape.measurements)
+    single_measure = num_measurements == 1
+    num_params = len(tape.trainable_params)
+    shot_vector = isinstance(shots, Sequence)
+    tape_specs = (single_measure, num_params, num_measurements, shot_vector, shots)
+
+    def processing_fn(results):
+        """Post-process the results of the parameter-shifted tapes of the pulse
+        generator parameter-shift rule into the gradient."""
+        grads = []
+        # Iterate over gradient_data, which is equivalent to the trainable parameters in argnum
+        for start, end, coeffs in gradient_data:
+            # If start and end pointer are equal, no tapes contribute and we get a vanishing
+            # gradient for this parameter. For this, add an entry ``None``, which will
+            # be replaced by the appropriately formatted zero later on.
+            if start == end:
+                grads.append(None)
+                continue
+
+            # Extract the results corresponding to the start and end pointer of the pulse
+            # to which the current parameter belongs.
+            res = results[start:end]
```

Comment:
Here is where the caching is happening, right? So if these generator terms have been evaluated before it should know where to get them?
 What I dont get it is how you can ensure that those have the same ordering as the `coeffs` from the current term? 🤔

Reply:
We extract potentially unnecessarily many results here: If a Pauli word occurs in one effective generator for the current pulse but not in the other, its derivative will always be extracted from the list of results. However, for the generator where it does _not_ appear, its derivative will be multiplied by `0` in the coefficients array. Does this make sense?

Reply:
resolved offline, problem was misunderstanding from my side thinking that `_generate_tapes_and_coeffs` treats every generator separately (which is not the case, they are all done in one pass through `_generate_tapes_and_coeffs` for one pulse gate.